src/js/storage/index.test.js

```diff
@@ -0,0 +1,51 @@
+import * as storage from './index';
+
+class LocalStorageMock {
+  constructor() {
+    this.value = {};
+  }
+
+  clear() {
+    this.value = {};
+  }
+
+  getItem(key) {
+    return this.value[key] || null;
+  }
+
+  setItem(key, value) {
+    this.value[key] = String(value);
+  }
+
+  removeItem(key) {
+    delete this.value[key];
+  }
+}
+
+global.localStorage = new LocalStorageMock();
+
+describe('Saving token to localstorage', () => {
+  it('Save token to storage', () => {
```

Comment:
You could adapt this to be your login/logout test

Reply:
```
require('dotenv/config');

import { login } from './login';
import * as storage from '../../storage/index.js';

// Correct login details
const userEmail = process.env.API_EMAIL;
const userPassword = process.env.API_PASSWORD;
const userToken = "gdfsgfdsgs2eysadfasaJhbGciOiJdfdsfdsfsdfwqrfIUdfdsfzI1NidfsfdsfIsInsdfdsfdsgfsdgR5c"
const data = { email: userEmail, password: userPassword };

// Mock login details
const mockEmail = "email@email.com"
const mockPassword = "1234"
const mockData = { email: mockEmail, password: mockPassword };
const mockResponse = {
  name: "fakeName",
  email: "fakeMail@fakeMail.xum",
  banner: null,
  avatar: null,
  accessToken: null
};

// Function to retrieve a success login
function fetchSuccess() {
  return Promise.resolve({
    ok: true,
    status: 201,
    statusText: 'All good! Take a nice coffee and relax',
    json: () => Promise.resolve(data),
  });
}

// Function to retrieve a fail login
function fetchFailure(status, statusText) {
  return Promise.resolve({
    ok: false,
    status,
    statusText,
  });
}

// Mock localstorage
class LocalStorageMock {
  constructor() {
    this.value = {};
  }

  clear() {
    this.value = {};
  }

  getItem(token) {
    return this.value[token] || null;
  }

  setItem(token, value) {
    this.value[token] = String(value);
  }

  removeItem(token) {
    delete this.value[token];
  }
}

global.localStorage = new LocalStorageMock();

describe('Authorization of user', () => {
  
  it('Login is a success', async () => {
    global.fetch = jest.fn(() => fetchSuccess());
    const response = await login(data);
    const token = storage.save("token", userToken)
    const profile = storage.save("profile", data)
    storage.save(token, profile);
    console.log(localStorage)
    expect(response)
  });
 
  it('Login is a failure', async () => {
    global.fetch = jest.fn(() => fetchFailure(401, 'Unauthorized'));
    await expect(login(mockData)).rejects.toThrow('Unauthorized');
  });
});

```
![image](https://user-images.githubusercontent.com/46559129/201094590-b849eb08-c3e3-4331-b685-6886f672dc89.png)
Not sure what this undefined is and where it comes from but it saves the correct data.

Reply:
Hmm this looks wrong to me `storage.save(token, profile);`

And you can't expect the response, since that's what you will get 100% of the time. You should expect a mock response object that you passed in to the login function

Reply:
Found my fault in the code. in the next commit and push it will be changed.